src/Nethermind/Nethermind.Serialization.Rlp/Rlp.cs

```diff
@@ -437,34 +437,47 @@ static void ThrowArgumentOutOfRangeException()
             }
         }
 
+        [SkipLocalsInit]
         public static Rlp Encode(ReadOnlySpan<byte> input)
         {
-            if (input.Length == 0)
+            // Handle special cases first
+            int length = input.Length;
+            if (length == 0)
             {
                 return OfEmptyByteArray;
             }
 
-            if (input.Length == 1 && input[0] < 128)
+            // If it's a single byte less than 128, it encodes to itself.
+            if (length == 1 && input[0] < 128)
             {
                 return new Rlp(input[0]);
             }
 
-            if (input.Length < 56)
+            // For lengths < 56, the encoding is one byte of prefix + the data
+            if (length < 56)
             {
-                byte smallPrefix = (byte)(input.Length + 128);
-                byte[] rlpResult = new byte[input.Length + 1];
-                rlpResult[0] = smallPrefix;
+                // Allocate exactly what we need: 1 prefix byte + input length
+                byte[] rlpResult = GC.AllocateUninitializedArray<byte>(1 + length);
+                // First byte is 0x80 + length
+                rlpResult[0] = (byte)(0x80 + length);
+                // Copy input after the prefix
                 input.CopyTo(rlpResult.AsSpan(1));
                 return new Rlp(rlpResult);
             }
             else
             {
-                byte[] serializedLength = SerializeLength(input.Length);
-                byte prefix = (byte)(183 + serializedLength.Length);
-                byte[] rlpResult = new byte[1 + serializedLength.Length + input.Length];
-                rlpResult[0] = prefix;
-                serializedLength.CopyTo(rlpResult.AsSpan(1));
-                input.CopyTo(rlpResult.AsSpan(1 + serializedLength.Length));
+                // For length >= 56, we need to encode the length itself (Big-Endian) and prepend a prefix
+                Span<byte> lengthBuffer = stackalloc byte[4]; // Enough for 32-bit lengths
+                int lengthOfLength = SerializeLength(length, lengthBuffer);
+                // Total size = 1 prefix byte + lengthOfLength + data length
+                int totalSize = 1 + lengthOfLength + length;
+                byte[] rlpResult = GC.AllocateUninitializedArray<byte>(totalSize);
```

Comment:
I think you could drop `lengthBuffer` and operate on `rlpResult` directly. It would need precalculation of  `lengthOfLength`

Reply:
👍 for @LukaszRozmej proposal. https://github.com/NethermindEth/Paprika/blob/5792d74fe1596265059aecd8b017f139403df305/src/Paprika/RLP/Rlp.cs#L59-L75 Count first, then serialize directly without additional copy of the length buffer

Reply:
Changed